tools/serve/serve.py

```diff
@@ -735,29 +734,30 @@ class Config(config.Config):
         "aliases": []
     }
 
+    computed_properties = ["ws_doc_root"] + config.ConfigBuilder.computed_properties
+
     def __init__(self, *args, **kwargs):
-        super(Config, self).__init__(
+        super(ConfigBuilder, self).__init__(
             subdomains=_subdomains,
             not_subdomains=_not_subdomains,
             *args,
             **kwargs
         )
 
-    @property
-    def ws_doc_root(self):
-        if self._ws_doc_root is not None:
-            return self._ws_doc_root
+    def _get_ws_doc_root(self, data):
```

Comment:
I'm confused by this; we appear to have a private getter never called from anywhere?

Reply:
OK, so what do you suggest? The requirements are
* Properties must be computed in a defined order, since later properties often depend on earlier ones.
* Subclasses must be able to add to the list of properties in a relatively straightforward manner.

We could just have a list of methods to call on the class, but that's a bit ugly because they have to be defined at the bottom of the class definition whereas it's more usual to define data properties at the top.

Reply:
A comment pointing at how it gets called and turned into an actual property would probably be useful?

Reply:
Added something to the docstring for the class describing the protocol.